src/go/github-runner-monitor/pkg/machines/machine.go

```diff
@@ -0,0 +1,194 @@
+		"-drive", fmt.Sprintf("id=vda,file=%s,aio=io_uring,format=qcow2,if=none", c.Disk), // spellcheck-ignore-line
+		"-snapshot",
+		// Create network explicitly. Use user-mode to avoid dealing with setup.
+		"-nic", "user,model=virtio-net-pci", // spellcheck-ignore-line
+		// Pass in qemu control socket; it's always 3 (after stdin/stdout/stderr).
+		"-qmp", "unix:fd=3,server=on",
+		// Set up VNC for debugging, but have it look for a free port automatically.
+		"-display", "none", "-vnc", ":0,to=9999",
+		// Expose GitHub runner configuration
+		"-fw_cfg", fmt.Sprintf("name=opt/github-jit-config,string=%s", c.JitConfig),
+	}
+
+	qmpLog := &LogAdapter{log}
+	chProcessDone := make(chan struct{})
+	procCtx, cancel := context.WithCancel(ctx)
+	created := false
+	attr := &unix.SysProcAttr{Setpgid: true} // Set PGID, so it doesn't receive Ctrl+C
+	cmd, stderr, err := qemu.LaunchCustomQemu(ctx, "", qemuArgs, []*os.File{qmpFile}, attr, qmpLog)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	qmpFile.Close() // Close the QMP server, since qemu has it now
+
+	// Signal when the process exits.
+	go func() {
+		if state, err := cmd.Process.Wait(); err != nil {
+			log.WithError(err).Error("Failed to wait for process.")
+		} else if !state.Success() {
+			log.Errorf("Process exited with %s", state.String())
+		}
+		close(chProcessDone)
+		cancel()
+	}()
+	// Copy the output from qemu to the logs in the background.
+	go func() {
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
+			log.Infof("%s", scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			log.WithError(err).Error("Error reading stderr")
+		}
+	}()
+
+	defer func() {
+		if !created {
+			killProcess(cmd, chProcessDone, log)
+		}
+	}()
+
+	// Now that the process is running, hook up the QMP channel.
+	conn, err := net.DialUnix("unix", nil, &net.UnixAddr{Name: qmpPath.Name()})
+	if err != nil {
+		return nil, fmt.Errorf("failed to establish connection for qmp: %w", err)
+	}
+
+	qmp, version, err := qemu.QMPStartWithConn(procCtx, conn, qemu.QMPConfig{Logger: qmpLog}, make(chan struct{}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to qmp: %w", err)
+	}
+	log.Tracef("Connected to QMP %d.%d.%d: %+v", version.Major, version.Minor, version.Micro, version.Capabilities)
+
+	if err = qmp.ExecuteQMPCapabilities(procCtx); err != nil {
+		log.WithError(err).Error("failed to get capabilities")
+		// Don't care about the failure other than to log it
+	}
+
+	// Set a custom function that will be called when procCtx is closed.
+	cmd.Cancel = func() error {
+		log.Info("Gracefully shutting down, this may take a while...")
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		if err := qmp.ExecuteSystemPowerdown(timeoutCtx); err != nil {
+			log.WithError(err).Error("Failed to power down")
+		}
+		cancel()
+		killProcess(cmd, chProcessDone, log)
```

Comment:
Should this also receive context?

Reply:
No, because on Ctrl+C, `ctx` would be closed, but we want `killProcess` to complete what it's doing.

(I initially had a context instead of a channel for `chProcessDone`, and realized that it was never shutting down gracefully.)